gce_workload_certs_refresh/main.go

```diff
@@ -0,0 +1,290 @@
+			return err
+		}
+		wtrc.RootCertificates[domain] = rc
+	}
+
+	return nil
+}
+
+// RootCertificate represents a Root Certificate in metadata
+type RootCertificate struct {
+	Metadata            Metadata
+	RootCertificatesPem string
+}
+
+// Metadata represents Metadata in metadata
+type Metadata struct {
+	WorkloadCredsDirPath string
+}
+
+func main() {
+	ctx := context.Background()
+
+	opts := logger.LogOpts{
+		LoggerName:     programName,
+		FormatFunction: logFormat,
+	}
+
+	opts.Writers = []io.Writer{os.Stderr}
+	logger.Init(ctx, opts)
+
+	// TODO: prune old dirs
+
+	if err := refreshCreds(); err != nil {
+		logger.Errorf(err.Error())
+	}
+
+	logger.Infof("Done")
+}
+
+func refreshCreds() error {
+	project, err := getMetadata("project/project-id")
+	if err != nil {
+		return fmt.Errorf("Error getting project ID: %v", err)
+	}
+	domain := fmt.Sprintf("%s.svc.id.goog", project)
+	logger.Infof("Rotating workload credentials for domain %s", domain)
+
+	wisMd, err := getMetadata("instance/workload-identities")
+	if err != nil {
+		return fmt.Errorf("Error getting workload-identities: %v", err)
+	}
+
+	wtrcsMd, err := getMetadata("instance/workload-trusted-root-certs")
+	if err != nil {
+		return fmt.Errorf("Error getting workload-identities: %v", err)
+	}
+
+	wis := WorkloadIdentities{}
+	if err := json.Unmarshal(wisMd, &wis); err != nil {
+		return fmt.Errorf("Error unmarshaling workload trusted root certs: %v", err)
+	}
+
+	wtrcs := WorkloadTrustedRootCerts{}
+	if err := json.Unmarshal(wtrcsMd, &wtrcs); err != nil {
+		return fmt.Errorf("Error unmarshaling workload trusted root certs: %v", err)
+	}
+
+	now := time.Now().Format(time.RFC3339)
+	contentDir := fmt.Sprintf("%s-%s", contentDirPrefix, now)
+	tempSymlink := fmt.Sprintf("%s-%s", tempSymlinkPrefix, now)
+
+	logger.Infof("Creating timestamp contents dir %s", contentDir)
+
+	// TODO: validate filesystem permissions
+	if err := os.MkdirAll(contentDir, 0750); err != nil {
+		return fmt.Errorf("Error creating contents dir: %v", err)
+	}
+
+	if err := os.WriteFile(fmt.Sprintf("%s/certificates.pem", contentDir), []byte(wis.WorkloadCredentials[domain].CertificatePem), 0666); err != nil {
+		return fmt.Errorf("Error writing certificates.pem: %v", err)
+	}
+
+	if err := os.WriteFile(fmt.Sprintf("%s/private_key.pem", contentDir), []byte(wis.WorkloadCredentials[domain].PrivateKeyPem), 0666); err != nil {
+		return fmt.Errorf("Error writing private_key.pem: %v", err)
+	}
+
+	if err := os.WriteFile(fmt.Sprintf("%s/ca_certificates.pem", contentDir), []byte(wtrcs.RootCertificates[domain].RootCertificatesPem), 0666); err != nil {
+		return fmt.Errorf("Error writing ca_certificates.pem: %v", err)
+	}
+
+	if err := os.Symlink(contentDir, tempSymlink); err != nil {
+		return fmt.Errorf("Error creating temporary link: %v", err)
+	}
+
+	logger.Infof("Rotating symlink %s", symlink)
+
+	if err := os.Rename(tempSymlink, symlink); err != nil {
+		return fmt.Errorf("Error rotating target link: %v", err)
```

Comment:
I'd suggest cleaning up the previous steps if failing to rotate the symlink.

Reply:
we intentionally want to leave the cruft for troubleshooting - a future PR will add some cleanup logic for very old things and possibly clean on subsequent success rather than on failure

Reply:
You mean you intend to implement a garbage collection logic? why are you later collecting garbage you could not have left?

Reply:
i usually think of the term 'garbage collection' as specific to in-memory structures. we want to leave the files present on failure for troubleshooting, because they will not interfere with normal operation - because we won't rotate the symlink.

Reply:
If you have a reason I'm fine with that - if these files left behind will be used for troubleshooting then ok - I'll not try to understand the kind of troubleshooting use that'd be :-)